### ISSUE TITLE
Add sources to bst shell

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#413]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/413)
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 30, 2018, 13:59

## Background

[//]: #
We want to be able to run software debuggers & co when we run `bst shell`

## Task description

[//]: # (Short summary of the action to be executed)

* [ ]  Modify `bst shell` to include the sources of all open workspaces in a given project
* [ ]  Potentially add a flag to enable/disable the feature (undecided)

## Steps

[//]: # (Acceptance criteria should follow the S.M.A.R.T. principle https://en.wikipedia.org/wiki/SMART_criteria )

* [ ]  Find the point were artifacts are added to the sandbox
* [ ]  Get a list of all open workspaces
* [ ]  Get the build-root for the elements of each workspace
* [ ]  Recreate the full path of each workspace inside the sandbox
    * This is to allow existing debug logs and paths to find the right path
    * In the case of conflicting paths between workspaces, the shell will abort and instruct the user to close some/all of the conflicting workspaces
* [ ]  Copy the contents of the workspaces into their corresponding directory 
* [ ]  Add tests to confirm that the files were copied into the sandbox

----

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on May 30, 2018, 14:33

Note that most of this functionality is probably covered in `Element.stage`, as elements are staged into the correct build root by default. I suspect this may be as simple as calling that method on every workspaced element before opening the shell, instead of modifying the staging process.

The only complication is that elements need to be staged into the build directory that the relevant artifact was built in, and not what the element thinks it should be built in, since debugging data in the artifact may refer to the original build directory. Since this could change between builds, we need to store more metadata - if this is not already covered.

We should also consider what happens if multiple elements stage into the same build directory. I have no real proposal for this, I believe the only option is to deduplicate them with `-n` suffixes and hope this doesn't break debuggers (with adequate warnings).

Note also that this feature should under no circumstances trigger on normal builds - this is purely debugging functionality.

If this is up for discussion, I personally think enabling this by default would be better, since the user can then find this feature through exploration, and debuggers will work out-of-the-box in most cases. There are no real downsides to staging a couple of workspaces, either.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 30, 2018, 15:03

> We should also consider what happens if multiple elements stage into the same build directory. I have no real proposal for this, I believe the only option is to deduplicate them with `-n` suffixes and hope this doesn't break debuggers (with adequate warnings). 

My recommendation would be that we abort, with a error message explaining why it failed and suggesting that, if they want to actually do this, they attempt to change the path of one of the listed elements that conflict

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 30, 2018, 15:04

However I think this would mean we can't have this enabled by default

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tlater]](https://gitlab.com/tlater) on May 30, 2018, 15:18

[[Gitlab user @knownexus]](https://gitlab.com/knownexus) if we do abort, we should suggest that the user closes all but one of the workspaces attempting to write to the same directory, instead of suggesting to change the directory (usually changing the default here means that it is required for the build to function, so the user simply could not use buildstream if we'd ask them to do that).

Otherwise, aborting seems like a fine solution. The advantage of deduplication is that the user could still ask their debugger to search in the correct directory, but I understand that this is not quite as intuitive. In either case, this should be an exceptionally rare use case, so I don't have a strong opinion.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 30, 2018, 15:24

Ok, in which case, i'll continue with the idea of abort and instruct the user to close some of the conflicting workspaces, and if it comes up as a problem/request later on, the deduplication functionality can be reconsidered

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 30, 2018, 15:26

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jun 8, 2018, 13:02

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jun 8, 2018, 13:02

marked the task **Find the point were artifacts are added to the sandbox** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jun 8, 2018, 13:03

marked the task **Get a list of all open workspaces** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jun 8, 2018, 13:03

marked the task **Recreate the full path of each workspace inside the sandbox** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jun 8, 2018, 13:03

marked the task **Copy the contents of the workspaces into their corresponding directory** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jun 8, 2018, 13:03

marked the task **Copy the contents of the workspaces into their corresponding directory** as incomplete

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jun 8, 2018, 13:03

marked the task **Recreate the full path of each workspace inside the sandbox** as incomplete

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jun 8, 2018, 14:41

marked the task **Get the build-root for the elements of each workspace** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jun 8, 2018, 14:41

marked the task **Recreate the full path of each workspace inside the sandbox** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jun 8, 2018, 14:41

marked the task **Copy the contents of the workspaces into their corresponding directory** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jun 11, 2018, 16:02

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) Thoughts on this plan?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jun 19, 2018, 17:41

* I think workspaces are mounted, not copied; if we were to have workspace presence in the shell environment, would this be any different ?
* Should we mount the build directories read-write in the sandbox ?
* This should also use cached build trees once they are implemented right ? How is this specific to workspaces ?
  * Is this issue just to handle the staging of build trees in `bst shell` where the edge case is that a dependency has an open workspace ?
  * Do we really know that mounting the workspaced dependency's workspace into `bst shell` is more correct than staging an existing build tree ?

Looks like there are a lot of choices to make, and we have to be careful not to expose too many options around this making things needlessly confusing:
* Which build directory is added to `bst shell`, the workspace or the cached built tree ?
* When staging the cached build trees of dependencies, the user will need to decide somewhere that it's time to download all these build trees, or to not bother with the exorbitant gigabytes of download and make due without the staged build trees.
  * Does the user get the option somehow to rebuild something locally, so as to avoid downloading a build tree artifact ? (I guess not)
* When a workspace exists on a dependency, does choosing the workspace depend on whether we have a cached build tree or not ?
  * I presume that if workspaces are used at all here; they are secondary to cached build trees; which are a much more certain representation of what was in the build tree at the time the artifact output was produced.

> My recommendation would be that we abort, with a error message explaining why it failed and suggesting that, if they want to actually do this, they attempt to change the path of one of the listed elements that conflict

For the case where multiple elements stage in the same dir, we could also handle this in the same way we handle overlap warning/errors.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jul 4, 2018, 12:14

This issue is currently being blocked by #455 and !474

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jul 4, 2018, 12:14

This issue is currently being blocked by #455 and !474

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Jul 4, 2018, 18:02

Moving back to ToDo and un-assigning

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jul 18, 2018, 11:18

unassigned [[Gitlab user @knownexus]](https://gitlab.com/knownexus)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Aug 7, 2018, 13:18

marked the task **Copy the contents of the workspaces into their corresponding directory** as incomplete

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Aug 7, 2018, 13:18

marked the task **Recreate the full path of each workspace inside the sandbox** as incomplete

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Aug 7, 2018, 13:18

marked the task **Get the build-root for the elements of each workspace** as incomplete

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Aug 7, 2018, 13:18

marked the task **Get a list of all open workspaces** as incomplete

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Aug 7, 2018, 13:18

marked the task **Get the build-root for the elements of each workspace** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Aug 7, 2018, 13:18

marked the task **Find the point were artifacts are added to the sandbox** as incomplete

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Aug 7, 2018, 13:19

marked the task **Find the point were artifacts are added to the sandbox** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Aug 7, 2018, 13:19

marked the task **Get the build-root for the elements of each workspace** as incomplete

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Aug 7, 2018, 13:19

marked the task **Find the point were artifacts are added to the sandbox** as incomplete

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Aug 8, 2018, 17:01

'Add sources to bst shell'

If you've built (or attempted to) $element & then you open a workspace for $element, then do 'bst shell --build $element' you'll be given a shell in which the source from the workspace is there (can test via file modification). This behaviour is also tested in integration/shell.py test_workspace_visible() & the same goes for mounting the sysroot into the shell. Am I missing something here?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Aug 8, 2018, 17:23

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @sstriker]](https://gitlab.com/sstriker) on Aug 9, 2018, 13:31

[[Gitlab user @tpollard]](https://gitlab.com/tpollard) the point is that any workspaces that are open for dependencies of the element, are not visible in the shell.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Aug 10, 2018, 06:20

I should point out that, while it *may* make sense (depending) to bind mount the *target* workspace while shelling into a build environment for that *target*, it does not make sense to me that we should bind mount all workspaces for the dependency elements.

Keep in mind that this feature is intended to improve debugging of a given build, and providing the source code for everything is a part of that, so that debuggers have access to the sources.

A workspace is outside of our control, and may have changed since the given build you want to debug occurred, as such you want the state of the workspace *at the time the build was performed*.

It does of course stand to reason that especially with `bst shell --build`, you will certainly want to bind mount the workspace of the *target* you are debugging the build of. If the element is already built and you are launching a runtime shell of an already built artifact, again I don't think you want to bind mount the workspace in that shell.

What makes most sense to me, is to use the cached *build tree* for all dependencies, regardless of whether those elements had open workspaces or not - with the only exception being a bind mount of the workspace of the *target*, when the invocation is a `bst shell --build` invocation.

### Some history / notes

I should note that we already behave properly for the case of debugging a failed build with either the `bst shell --sysroot` option, or from the interactive prompt.

Since d46fbbcbfec7f44cfa3d412b70000eff633b85bb (a fix for regression #346), we are making a copy of the workspace state at the time of failure, ensuring that we are debugging exactly what failed and not the arbitrary workspace state when later debugging the failed build.

Additionally I will note that this functionality should be replaced with the new build tree caching mechanism, this should be done implicitly while ensuring we only have one codepath to debug a failed build, which is an objective outlined in #539.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Aug 22, 2018, 12:03

marked this issue as related to #432

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @bethwhite]](https://gitlab.com/bethwhite) on Sep 3, 2018, 13:51

Addressed in the workspaces proposal on the mailing list [here][1]

[1]: https://mail.gnome.org/archives/buildstream-list/2018-August/msg00078.html
